asdf/resource.py

```diff
@@ -0,0 +1,180 @@
+"""
+Support for plugins that provide access to resources such
+as schemas.
+"""
+from collections.abc import Mapping
+from pathlib import Path
+import fnmatch
+import os
+import pkgutil
+import sys
+
+if sys.version_info < (3, 9):
+    import importlib_resources
+else:
+    import importlib.resources as importlib_resources
+
+import asdf
+
+
+__all__ = ["TraversableResourceMapping", "ResourceManager"]
+
+
+class TraversableResourceMapping(Mapping):
+    """
+    Resource mapping that reads resource content
+    from a root directory.
+
+    Parameters
+    ----------
+    root : str or importlib.resources.abc.Traversable
+        Root directory of the resource files.  `str` will
+        be interpreted as a filesystem path.
+    uri_prefix : str
+        Prefix used to construct URIs from file paths.  The
+        prefix will be prepended to paths relative to the root
+        directory.
+    recursive : bool, optional
+        If `True`, recurse into subdirectories.  Defaults to `False`.
+    filename_pattern : str, optional
+        Glob pattern that identifies relevant filenames.
+        Defaults to "*.yaml".
+    stem_filename : bool, optional
+        If `True`, remove the filename's extension when
+        constructing its URI.
+    """
+    def __init__(self, root, uri_prefix, recursive=False, filename_pattern="*.yaml", stem_filename=True):
+        self._uri_to_traversable = {}
+        self._recursive = recursive
+        self._filename_pattern = filename_pattern
+        self._stem_filename = stem_filename
+
+        if isinstance(root, str):
+            self._root = Path(root)
+        else:
+            self._root = root
+
+        if uri_prefix.endswith("/"):
+            self._uri_prefix = uri_prefix[:-1]
+        else:
+            self._uri_prefix = uri_prefix
+
+        for traversable, path in self._iterate_files(self._root, []):
```

Comment:
I find the use of the name traversable when a file (or equivalent is returned) misleading. Shouldn't it be something based on file? (name, that is)?

Reply:
I think it would be fair to change this from `traversable` to `file` (and maybe the name of the class from `TraversableResourceMapping` to `DirectoryResourceMapping`.  I was trying to show that it works with any implementation of [importlib.abc.Traversable](https://github.com/python/cpython/blob/3.9/Lib/importlib/abc.py#L393), but that's a bit esoteric given that this interface is introduced in the yet-unreleased Python 3.9.  I can document the possible use of a non-directory object in one of the docstrings.

Reply:
Changed the name to `DirectoryResourceMapping`, and added a test that confirms that the class still works with a non-pathlib Traversable.